### PR TITLE
Mock email lookup response if in sandbox environment [ECCT-837]

### PR DESCRIPTION
**Pull request by nevans4-ch**
Bean to implement interface PrivateDataRetrievalService is instantiated depending on environment: OracleQueryApiDataRetrievalServiceImpl for normal operation, MockEmailRetrieverImpl if in a sandbox environment.

MockEmailRetrieverImpl returns a mocked successful email address lookup if the company number ends with ERR, null otherwise.  This results in the sandbox environments responding to the eligibility check endpoint with {"eligibility_status_code":"COMPANY_VALID_FOR_SERVICE"} for ERR companies, and {"eligibility_status_code":"INVALID_NO_REGISTERED_EMAIL_ADDRESS_EXISTS"} for all others (if companies pass other eligibility checks).

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/registeredemailaddressapi/service/MockEmailRetrieverImpl.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression;
+import org.springframework.stereotype.Service;
+import uk.gov.companieshouse.api.model.company.RegisteredEmailAddressJson;
+import uk.gov.companieshouse.registeredemailaddressapi.exception.ServiceException;
+import uk.gov.companieshouse.registeredemailaddressapi.utils.ApiLogger;
+
+import java.util.HashMap;
+
+@Service
+@Qualifier("mockEmailRetrieverImpl")
```
Comment (pparlett) [inline src/main/java/uk/gov/companieshouse/registeredemailaddressapi/service/MockEmailRetrieverImpl.java:14]: this qualifier is not referenced

File: src/main/java/uk/gov/companieshouse/registeredemailaddressapi/service/MockEmailRetrieverImpl.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+@ConditionalOnExpression("'${env.name}'.equals('stagsbox') || '${env.name}'.equals('livesbox')")
```
Comment (pparlett) [inline src/main/java/uk/gov/companieshouse/registeredemailaddressapi/service/MockEmailRetrieverImpl.java:15]: I understand how this works in conjunction with the conditional in the Oracle impl, pity it is not possible to have one as "default" (the Oracle one) and other (this mock) as an override

File: src/main/java/uk/gov/companieshouse/registeredemailaddressapi/service/MockEmailRetrieverImpl.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+public class MockEmailRetrieverImpl implements PrivateDataRetrievalService {
+
+    @Override
+    public RegisteredEmailAddressJson getRegisteredEmailAddress(String companyNumber) throws ServiceException {
+
+        var logMap = new HashMap<String, Object>();
+
+        if (companyNumber.endsWith("ERR")) {
```
Comment (pparlett) [inline src/main/java/uk/gov/companieshouse/registeredemailaddressapi/service/MockEmailRetrieverImpl.java:23]: may be a constant ERR for extra clarity

File: src/main/java/uk/gov/companieshouse/registeredemailaddressapi/service/MockEmailRetrieverImpl.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+            ApiLogger.info(format("Mocking Registered Email Address lookup for sandbox environment - returning mock email address for company number %s", companyNumber), logMap);
+            var registeredEmailAddressJson = new RegisteredEmailAddressJson();
+            registeredEmailAddressJson.setRegisteredEmailAddress("mockexistingemail@companieshouse.gov.uk");
```
Comment (pparlett) [inline src/main/java/uk/gov/companieshouse/registeredemailaddressapi/service/MockEmailRetrieverImpl.java:26]: may be a constant MOCK_EMAIL for extra clarity

File: src/main/java/uk/gov/companieshouse/registeredemailaddressapi/service/PrivateDataRetrievalService.java [hunk 5]
```diff
@@ -1,62 +1,11 @@
...
-                    .getCompanyRegisteredEmailAddress(String.format(REGISTERED_EMAIL_ADDRESS_URI_SUFFIX, companyNumber))
-                    .execute()
-                    .getData();
-
-            ApiLogger.info("Successfully retrieved Registered Email Address from database", logMap);
-            return registeredEmailAddressJson;
 
-        } catch (ApiErrorResponseException e) {
-            ApiLogger.error("Error Retrieving Registered Email Address for Company ", e, logMap);
+@Component
```
Comment (pparlett) [inline src/main/java/uk/gov/companieshouse/registeredemailaddressapi/service/PrivateDataRetrievalService.java:7]: harmless but unnecessary

File: src/main/java/uk/gov/companieshouse/registeredemailaddressapi/service/RegisteredEmailAddressService.java [hunk 6]
```diff
@@ -264,7 +264,3 @@ private void updateTransactionWithLinks(Transaction transaction,
     }
 
 }
-
```
Comment (pparlett) [inline src/main/java/uk/gov/companieshouse/registeredemailaddressapi/service/RegisteredEmailAddressService.java:267]: I was wondering how this file got included in the change set ;-)

File: src/test/java/uk/gov/companieshouse/registeredemailaddressapi/integration/EligibilityControllerEmailMockingIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@SpringBootTest
+@AutoConfigureMockMvc
+@TestPropertySource(properties = {"env.name = livesbox"})
+class EligibilityControllerEmailMockingIntegrationTest {
+
+    Helper helper = new Helper();
```
Comment (pparlett) [inline src/test/java/uk/gov/companieshouse/registeredemailaddressapi/integration/EligibilityControllerEmailMockingIntegrationTest.java:25]: missing access specifier; could be "private static final HELPER = ..." and other occurrences; (in fact the Helper should be fixed such that all methods are statis and the class made final with a private constructor - but may be not your problem!)

File: src/test/java/uk/gov/companieshouse/registeredemailaddressapi/integration/EligibilityControllerEmailMockingIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    @MockBean
+    protected CompanyProfileService companyProfileService;
+
+
+    @Test
+    void EligibilityEndpointMockEmailResponseTest() throws Exception {
+
+        final String companyNumber = "12345ERR";
+        CompanyProfileApi companyProfileApi = helper.generateCompanyProfileApi(companyNumber);
+        when(companyProfileService.getCompanyProfile(companyNumber)).thenReturn(companyProfileApi);
```
Comment (pparlett) [inline src/test/java/uk/gov/companieshouse/registeredemailaddressapi/integration/EligibilityControllerEmailMockingIntegrationTest.java:39]: surprised you have not taken up the given().willReturn() style in BDD for new tests or may be you prefer consistency with existing tests

File: src/test/java/uk/gov/companieshouse/registeredemailaddressapi/integration/EligibilityControllerEmailMockingIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+@TestPropertySource(properties = {"env.name = livesbox"})
+class EligibilityControllerEmailMockingIntegrationTest {
+
+    Helper helper = new Helper();
+
+    @Autowired
+    private MockMvc mvc;
+
+    @MockBean
+    protected CompanyProfileService companyProfileService;
```
Comment (pparlett) [inline src/test/java/uk/gov/companieshouse/registeredemailaddressapi/integration/EligibilityControllerEmailMockingIntegrationTest.java:31]: why is this protected and not private?

File: src/test/java/uk/gov/companieshouse/registeredemailaddressapi/integration/RegisteredEmailAddressControllerIntegrationTest.java [hunk 10]
```diff
@@ -49,6 +50,7 @@ class RegisteredEmailAddressControllerIntegrationTest {
     protected TransactionService transactionService;
 
     @MockBean
+    @Qualifier("oracleQueryApiDataRetrievalServiceImpl")
     protected PrivateDataRetrievalService privateDataRetrievalService;
```
Comment (pparlett) [inline src/test/java/uk/gov/companieshouse/registeredemailaddressapi/integration/RegisteredEmailAddressControllerIntegrationTest.java:54]: (why protected? etc other comments as per new test, but not your problem for this existing test)

File: src/test/java/uk/gov/companieshouse/registeredemailaddressapi/unit/service/MockEmailRetrieverImplTest.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+    private static final String COMPANY_NUMBER = "12345678";
+    private static final String COMPANY_NUMBER_ERR = "12345ERR";
+    private static final String COMPANY_EMAIL_MOCK = "mockexistingemail@companieshouse.gov.uk";
+
+
+    private MockEmailRetrieverImpl mockEmailRetrieverImpl = new MockEmailRetrieverImpl();
+
+
+    @Nested
```
Comment (pparlett) [inline src/test/java/uk/gov/companieshouse/registeredemailaddressapi/unit/service/MockEmailRetrieverImplTest.java:21]: cannot see the need for this, same comment for the Oracle impl test class

File: src/test/java/uk/gov/companieshouse/registeredemailaddressapi/unit/service/OracleQueryApiDataRetrievalServiceImplTest.java [hunk 12]
```diff
@@ -24,16 +24,16 @@
...
 @ExtendWith(MockitoExtension.class)
-class PrivateDataRetrievalServiceTest {
+class OracleQueryApiDataRetrievalServiceImplTest {
 
     private static final String COMPANY_NUMBER = "12345678";
     private static final String COMPANY_EMAIL = "tester@testing.com";
 
     @InjectMocks
-    private PrivateDataRetrievalService privateDataRetrievalService;
+    private OracleQueryApiDataRetrievalServiceImpl oracleQueryApiDataRetrievalServiceImpl;
```
Comment (pparlett) [inline src/test/java/uk/gov/companieshouse/registeredemailaddressapi/unit/service/OracleQueryApiDataRetrievalServiceImplTest.java:36]: line 54 has constant with incorrect naming convention

File: src/test/resources/application.properties [hunk 13]
```diff
@@ -4,3 +4,4 @@ REGISTERED_EMAIL_ADDRESS_UPDATE_FILING_DESCRIPTION=Registered Email Address upda
 
 allowed.company.statuses = active, liquidation, receivership, voluntary-agreement, insolvency-proceedings, administration
 allowed.company.types = private-unlimited, ltd, plc, private-limited-guarant-nsc-limited-exemption, private-limited-guarant-nsc, private-unlimited-nsc, private-limited-shares-section-30-exemption, llp
+env.name=test
```
Comment (pparlett) [inline src/test/resources/application.properties:7]: is this needed?

Comment (nevans4-ch) [reply src/test/resources/application.properties:7]: Good spot - no, it was needed for a previous version of this during development but not any more.

File: src/test/java/uk/gov/companieshouse/registeredemailaddressapi/unit/service/MockEmailRetrieverImplTest.java (hunk 11 above)
Comment (nevans4-ch) [reply src/test/java/uk/gov/companieshouse/registeredemailaddressapi/unit/service/MockEmailRetrieverImplTest.java:21]: The existing (renamed in this PR) Oracle impl test class had this nesting first, this parallel one is just following the same structure.  In general I like the clarity added to test output by this kind of grouping, but agree that it isn't needed here as neither has more than one group of tests.  I am inclined to leave it here as it isn't doing any harm, unless you feel strongly?

Comment (pparlett) [reply src/test/java/uk/gov/companieshouse/registeredemailaddressapi/unit/service/MockEmailRetrieverImplTest.java:21]: As per your preference

**Review by pparlett — APPROVED** [review]
Looks good

**Pull request merged**